File: services/quant-engine/app/services/construction_policy_catalog.py
from __future__ import annotations

from dataclasses import dataclass
from fractions import Fraction
from typing import Callable

from app.schemas.construction import (
    ConstructionPolicyCatalogEntry,
    ConstructionPolicyId,
    ConstructionPolicyLaunchProfile,
    ConstructionPolicyLaunchProfilePolicyStatus,
    ConstructionRankedCandidateInput,
    ConstructionSelectionRuleId,
    ConstructionWeight,
)

ELIGIBLE_ONLY_RULE_ID: ConstructionSelectionRuleId = "eligible_only"
TAKE_TOP_N_RULE_ID: ConstructionSelectionRuleId = "take_top_n"
RANKING_ARTIFACT_REVIEW_HANDOFF_PROFILE_ID = "ranking_artifact_review_handoff_v1"


def _ranking_artifact_review_handoff_launch_profile(
    policy_status: ConstructionPolicyLaunchProfilePolicyStatus,
) -> ConstructionPolicyLaunchProfile:
    return ConstructionPolicyLaunchProfile(
        profile_id=RANKING_ARTIFACT_REVIEW_HANDOFF_PROFILE_ID,
        profile_kind="ranking_artifact_review_handoff",
        policy_status=policy_status,
        launch_top_n=2,
    )


def _validate_launch_profile(definition: ConstructionPolicyDefinition) -> None:
    entry = definition.catalog_entry
    profile = entry.launch_profile
    if profile.profile_id != RANKING_ARTIFACT_REVIEW_HANDOFF_PROFILE_ID:
        raise ValueError(f"construction policy {entry.policy_id} has unsupported launch_profile.profile_id")
    if profile.profile_kind != "ranking_artifact_review_handoff":
        raise ValueError(f"construction policy {entry.policy_id} has unsupported launch_profile.profile_kind")
    if profile.launch_top_n != entry.launch_top_n:
        raise ValueError(f"construction policy {entry.policy_id} launch profile launch_top_n must match launch_top_n")

    expected_rows = {
        "top_n_equal_weight_v1": {
            "policy_definition_id": "construction_policy_definition_top_n_equal_weight_v1",
            "ranking_support": "selection_only",
            "launch_status": "default",
        },
        "top_n_linear_rank_weight_v1": {
            "policy_definition_id": "construction_policy_definition_top_n_linear_rank_weight_v1",
            "ranking_support": "linear_selected_order_weighting",
            "launch_status": "opt_in",
        },
        "top_n_inverse_rank_weight_v1": {
            "policy_definition_id": "construction_policy_definition_top_n_inverse_rank_weight_v1",
            "ranking_support": "inverse_selected_order_weighting",
            "launch_status": "excluded",
        },
    }
    expected = expected_rows.get(entry.policy_id)
    if expected is None:
        raise ValueError(f"construction policy {entry.policy_id} is not supported by the canonical launch profile")
    if entry.policy_definition_id != expected["policy_definition_id"]:
        raise ValueError(f"construction policy {entry.policy_id} launch profile metadata disagrees with policy_definition_id")
    if entry.ranking_support != expected["ranking_support"]:
        raise ValueError(f"construction policy {entry.policy_id} launch profile metadata disagrees with ranking_support")
    if profile.policy_status != expected["launch_status"]:
        raise ValueError(f"construction policy {entry.policy_id} launch profile metadata disagrees with policy_id")


def _validate_canonical_launch_profile_catalog(policy_catalog: tuple[ConstructionPolicyDefinition, ...]) -> None:
    matching_profiles = [
        definition
        for definition in policy_catalog
        if definition.catalog_entry.launch_profile.profile_id == RANKING_ARTIFACT_REVIEW_HANDOFF_PROFILE_ID
    ]
    if not matching_profiles:
        raise ValueError("construction policy catalog is missing the canonical ranking-artifact review handoff launch profile")

    default_rows = [
        definition.catalog_entry.policy_id
        for definition in matching_profiles
        if definition.catalog_entry.launch_profile.policy_status == "default"
    ]
    if len(default_rows) != 1 or default_rows[0] != "top_n_equal_weight_v1":
        raise ValueError("construction policy catalog must define exactly one default launch policy for ranking-artifact review handoff")

    included_rows = sorted(
        definition.catalog_entry.policy_id
        for definition in matching_profiles
        if definition.catalog_entry.launch_profile.policy_status in {"default", "opt_in"}
    )
    if included_rows != ["top_n_equal_weight_v1", "top_n_linear_rank_weight_v1"]:
        raise ValueError("construction policy catalog ranking-artifact review handoff profile must include only equal-weight and linear-rank launch policies")

    excluded_rows = sorted(
        definition.catalog_entry.policy_id
        for definition in matching_profiles
        if definition.catalog_entry.launch_profile.policy_status == "excluded"
    )
    if excluded_rows != ["top_n_inverse_rank_weight_v1"]:
        raise ValueError("construction policy catalog ranking-artifact review handoff profile must exclude inverse-rank from launch")

    for definition in matching_profiles:
        _validate_launch_profile(definition)


@dataclass(frozen=True)
class ConstructionPolicyDefinition:
    catalog_entry: ConstructionPolicyCatalogEntry
    max_position_failure_reason: str
    cutoff_exclusion_reason: str
    raw_weight_numerator_builder: Callable[[int], list[Fraction]]

    @property
    def selection_rule_ids(self) -> tuple[ConstructionSelectionRuleId, ...]:
        return tuple(self.catalog_entry.selection_rule_ids)


def _equal_weight_numerators(selected_count: int) -> list[Fraction]:
    return [Fraction(1, selected_count)] * selected_count


def _inverse_rank_weight_numerators(selected_count: int) -> list[Fraction]:
    return [Fraction(1, selected_order_rank) for selected_order_rank in range(1, selected_count + 1)]


def _linear_rank_weight_numerators(selected_count: int) -> list[Fraction]:
    return [Fraction(selected_count - selected_order_rank + 1, 1) for selected_order_rank in range(1, selected_count + 1)]


POLICY_CATALOG: tuple[ConstructionPolicyDefinition, ...] = (
    ConstructionPolicyDefinition(
        catalog_entry=ConstructionPolicyCatalogEntry(
            policy_id="top_n_equal_weight_v1",
            policy_definition_id="construction_policy_definition_top_n_equal_weight_v1",
            name="Top N Equal Weight v1",
            description="Select eligible top-ranked names and assign equal target weights.",
            family="top_n_equal_weight",
            constraints="long_only_fully_invested_max_position_turnover",
            inputs="ranked_universe_and_current_portfolio",
            determinism="deterministic_rank_order",
            ranking_support="selection_only",
            full_investment_constraint="required",
            long_only_constraint="required",
            eligible_ranked_universe_constraint="required",
            max_position_weight_constraint="required",
            min_position_weight_constraint="supported_optional",
            max_turnover_weight_constraint="supported_optional",
            max_trade_intent_count_constraint="supported_optional",
            ranked_universe_input="required",
            current_portfolio_input="required",
            launch_top_n=2,
            selection_rule_ids=[ELIGIBLE_ONLY_RULE_ID, TAKE_TOP_N_RULE_ID],
            launch_profile=_ranking_artifact_review_handoff_launch_profile("default"),
        ),
        max_position_failure_reason="equal-weight seed exceeds max_position_weight",
        cutoff_exclusion_reason="not selected by top_n_equal_weight_v1 cutoff",
        raw_weight_numerator_builder=_equal_weight_numerators,
    ),
    ConstructionPolicyDefinition(
        catalog_entry=ConstructionPolicyCatalogEntry(
            policy_id="top_n_inverse_rank_weight_v1",
            policy_definition_id="construction_policy_definition_top_n_inverse_rank_weight_v1",
            name="Top N Inverse Rank Weight v1",
            description="Select eligible top-ranked names and weight them by inverse selected-order rank.",
            family="top_n_rank_weighted",
            constraints="long_only_fully_invested_max_position_turnover",
            inputs="ranked_universe_and_current_portfolio",
            determinism="deterministic_rank_order",
            ranking_support="inverse_selected_order_weighting",
            full_investment_constraint="required",
            long_only_constraint="required",
            eligible_ranked_universe_constraint="required",
            max_position_weight_constraint="required",
            min_position_weight_constraint="supported_optional",
            max_turnover_weight_constraint="supported_optional",
            max_trade_intent_count_constraint="supported_optional",
            ranked_universe_input="required",
            current_portfolio_input="required",
            launch_top_n=2,
            selection_rule_ids=[ELIGIBLE_ONLY_RULE_ID, TAKE_TOP_N_RULE_ID],
            launch_profile=_ranking_artifact_review_handoff_launch_profile("excluded"),
        ),
        max_position_failure_reason="inverse-rank seed exceeds max_position_weight",
        cutoff_exclusion_reason="not selected by top_n_inverse_rank_weight_v1 cutoff",
        raw_weight_numerator_builder=_inverse_rank_weight_numerators,
    ),
    ConstructionPolicyDefinition(
        catalog_entry=ConstructionPolicyCatalogEntry(
            policy_id="top_n_linear_rank_weight_v1",
            policy_definition_id="construction_policy_definition_top_n_linear_rank_weight_v1",
            name="Top N Linear Rank Weight v1",
            description="Select eligible top-ranked names and weight them by selected-order linear rank numerators N..1.",
            family="top_n_rank_weighted",
            constraints="long_only_fully_invested_max_position_turnover",
            inputs="ranked_universe_and_current_portfolio",
            determinism="deterministic_rank_order",
            ranking_support="linear_selected_order_weighting",
            full_investment_constraint="required",
            long_only_constraint="required",
            eligible_ranked_universe_constraint="required",
            max_position_weight_constraint="required",
            min_position_weight_constraint="supported_optional",
            max_turnover_weight_constraint="supported_optional",
            max_trade_intent_count_constraint="supported_optional",
            ranked_universe_input="required",
            current_portfolio_input="required",
            launch_top_n=2,
            selection_rule_ids=[ELIGIBLE_ONLY_RULE_ID, TAKE_TOP_N_RULE_ID],
            launch_profile=_ranking_artifact_review_handoff_launch_profile("opt_in"),
        ),
        max_position_failure_reason="linear-rank seed exceeds max_position_weight",
        cutoff_exclusion_reason="not selected by top_n_linear_rank_weight_v1 cutoff",
        raw_weight_numerator_builder=_linear_rank_weight_numerators,
    ),
)

_POLICY_BY_ID: dict[str, ConstructionPolicyDefinition] = {
    definition.catalog_entry.policy_id: definition for definition in POLICY_CATALOG
}

_validate_canonical_launch_profile_catalog(POLICY_CATALOG)


def list_construction_policies(
    *,
    family: str | None = None,
    constraints: str | None = None,
    inputs: str | None = None,
    determinism: str | None = None,
    ranking_support: str | None = None,
    full_investment_constraint: str | None = None,
    long_only_constraint: str | None = None,
    eligible_ranked_universe_constraint: str | None = None,
    max_position_weight_constraint: str | None = None,
    min_position_weight_constraint: str | None = None,
    max_turnover_weight_constraint: str | None = None,
    max_trade_intent_count_constraint: str | None = None,
    ranked_universe_input: str | None = None,
    current_portfolio_input: str | None = None,
    launch_top_n: int | None = None,
) -> list[ConstructionPolicyCatalogEntry]:
    _validate_canonical_launch_profile_catalog(POLICY_CATALOG)
    policies: list[ConstructionPolicyCatalogEntry] = []
    for definition in POLICY_CATALOG:
        entry = definition.catalog_entry
        if family is not None and entry.family != family:
            continue
        if constraints is not None and entry.constraints != constraints:
            continue
        if inputs is not None and entry.inputs != inputs:
            continue
        if determinism is not None and entry.determinism != determinism:
            continue
        if ranking_support is not None and entry.ranking_support != ranking_support:
            continue
        if full_investment_constraint is not None and entry.full_investment_constraint != full_investment_constraint:
            continue
        if long_only_constraint is not None and entry.long_only_constraint != long_only_constraint:
            continue
        if (
            eligible_ranked_universe_constraint is not None
            and entry.eligible_ranked_universe_constraint != eligible_ranked_universe_constraint
        ):
            continue
        if max_position_weight_constraint is not None and entry.max_position_weight_constraint != max_position_weight_constraint:
            continue
        if min_position_weight_constraint is not None and entry.min_position_weight_constraint != min_position_weight_constraint:
            continue
        if max_turnover_weight_constraint is not None and entry.max_turnover_weight_constraint != max_turnover_weight_constraint:
            continue
        if (
            max_trade_intent_count_constraint is not None
            and entry.max_trade_intent_count_constraint != max_trade_intent_count_constraint
        ):
            continue
        if ranked_universe_input is not None and entry.ranked_universe_input != ranked_universe_input:
            continue
        if current_portfolio_input is not None and entry.current_portfolio_input != current_portfolio_input:
            continue
        if launch_top_n is not None and entry.launch_top_n != launch_top_n:
            continue
        policies.append(entry.model_copy(deep=True))
    return policies


def get_construction_policy_definition(policy_id: str) -> ConstructionPolicyDefinition | None:
    return _POLICY_BY_ID.get(policy_id)


def build_policy_weights(
    definition: ConstructionPolicyDefinition,
    selected: list[ConstructionRankedCandidateInput],
    *,
    max_position_weight: float,
    epsilon: float,
    normalize_weights,
) -> tuple[list[ConstructionWeight], list[ConstructionWeight], str | None]:
    if not selected:
        return [], [], None
    weights = normalize_weights(definition.raw_weight_numerator_builder(len(selected)))
    failure_reason = None
    if weights and max(weights) > max_position_weight + epsilon:
        failure_reason = definition.max_position_failure_reason
    construction_weights = [
        ConstructionWeight(symbol=item.symbol, weight=weight)
        for item, weight in zip(selected, weights, strict=True)
    ]
    return construction_weights, construction_weights.copy(), failure_reason


def get_policy_cutoff_exclusion_reason(definition: ConstructionPolicyDefinition) -> str:
    return definition.cutoff_exclusion_reason

File: apps/desktop/src/features/backtest/constructionPolicyCatalog.ts
import { useEffect, useMemo, useState } from 'react'

import type {
  ConstructionPolicyLaunchProfile,
  ConstructionDiscoveredPolicy,
  ConstructionPolicyRunInput,
} from '../portfolio/types'

const DESKTOP_CONSTRUCTION_POLICY_FILTERS = {
  constraints: 'long_only_fully_invested_max_position_turnover',
  inputs: 'ranked_universe_and_current_portfolio',
  determinism: 'deterministic_rank_order',
  full_investment_constraint: 'required',
  long_only_constraint: 'required',
  eligible_ranked_universe_constraint: 'required',
  max_position_weight_constraint: 'required',
  min_position_weight_constraint: 'supported_optional',
  max_turnover_weight_constraint: 'supported_optional',
  max_trade_intent_count_constraint: 'supported_optional',
  ranked_universe_input: 'required',
  current_portfolio_input: 'required',
  launch_top_n: '2',
} as const

// Epic 3 breadth: launch top_n widened from a hardcoded 2 to a configurable range.
// The DEFAULT value remains 2 for backward compatibility — users can pick any value
// in the supported range via the construction browser UI.
export const RANKING_ARTIFACT_CONSTRUCTION_LAUNCH_TOP_N_DEFAULT = 2 as const
export const RANKING_ARTIFACT_CONSTRUCTION_LAUNCH_TOP_N_MIN = 2 as const
export const RANKING_ARTIFACT_CONSTRUCTION_LAUNCH_TOP_N_MAX = 20 as const
/** @deprecated Use RANKING_ARTIFACT_CONSTRUCTION_LAUNCH_TOP_N_DEFAULT (default value)
 *  or one of the _MIN/_MAX bounds depending on context. Kept as alias for callers
 *  that still treat top_n as a fixed scalar — they get the default value. */
export const RANKING_ARTIFACT_CONSTRUCTION_LAUNCH_TOP_N = RANKING_ARTIFACT_CONSTRUCTION_LAUNCH_TOP_N_DEFAULT

const SUPPORTED_POLICY_IDS = new Set([
  'top_n_equal_weight_v1',
  'top_n_inverse_rank_weight_v1',
  'top_n_linear_rank_weight_v1',
])

const SUPPORTED_POLICY_SPECS = {
  top_n_equal_weight_v1: {
    policy_definition_id: 'construction_policy_definition_top_n_equal_weight_v1',
    family: 'top_n_equal_weight',
    ranking_support: 'selection_only',
  },
  top_n_inverse_rank_weight_v1: {
    policy_definition_id: 'construction_policy_definition_top_n_inverse_rank_weight_v1',
    family: 'top_n_rank_weighted',
    ranking_support: 'inverse_selected_order_weighting',
  },
  top_n_linear_rank_weight_v1: {
    policy_definition_id: 'construction_policy_definition_top_n_linear_rank_weight_v1',
    family: 'top_n_rank_weighted',
    ranking_support: 'linear_selected_order_weighting',
  },
} as const

const CANONICAL_LAUNCH_PROFILE_ID = 'ranking_artifact_review_handoff_v1'
const CANONICAL_LAUNCH_PROFILE_KIND = 'ranking_artifact_review_handoff'

const CANONICAL_LAUNCH_POLICY_STATUS = {
  top_n_equal_weight_v1: 'default',
  top_n_inverse_rank_weight_v1: 'excluded',
  top_n_linear_rank_weight_v1: 'opt_in',
} as const satisfies Record<keyof typeof SUPPORTED_POLICY_SPECS, ConstructionPolicyLaunchProfile['policy_status']>

const CANONICAL_LAUNCH_INCLUDED_POLICY_IDS = new Set([
  'top_n_equal_weight_v1',
  'top_n_linear_rank_weight_v1',
])

export function resolvePolicyDefinitionIdForPolicyId(policyId: string): string | null {
  return SUPPORTED_POLICY_SPECS[policyId as keyof typeof SUPPORTED_POLICY_SPECS]?.policy_definition_id ?? null
}

function isLaunchIncludedPolicy(policy: ConstructionDiscoveredPolicy): boolean {
  return policy.launch_profile.policy_status === 'default' || policy.launch_profile.policy_status === 'opt_in'
}

const SUPPORTED_RANKING_SUPPORT = new Set([
  'selection_only',
  'inverse_selected_order_weighting',
  'linear_selected_order_weighting',
])

const REQUIRED_SELECTION_RULE_IDS = ['eligible_only', 'take_top_n'] as const

type ConstructionPolicyCatalogState = {
  status: 'idle' | 'loading' | 'ready' | 'error'
  policies: ConstructionDiscoveredPolicy[]
  error: string | null
}

export type ConstructionLaunchPolicyReadback = {
  policyName: string
  statusLabel: 'default' | 'opt-in'
  topN: 2
  requiredConstraint: 'max_position_weight'
  optionalConstraints: Array<'min_position_weight' | 'max_turnover_weight' | 'max_trade_intent_count'>
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value != null
}

async function readJsonResponse<T>(response: Response, fallbackMessage: string) {
  const payload = await response.json().catch(() => null)
  if (!response.ok) {
    throw new Error(typeof payload === 'object' && payload != null && 'detail' in payload && typeof payload.detail === 'string' ? payload.detail : fallbackMessage)
  }
  return payload as T
}

function validateExactField<T extends string>(
  row: Record<string, unknown>,
  field: string,
  expected: T,
): T {
  const value = row[field]
  if (value !== expected) {
    throw new Error(`Construction policy catalog row has unsupported ${field}`)
  }
  return value as T
}

function parseLaunchProfile(payload: unknown, policyId: keyof typeof SUPPORTED_POLICY_SPECS): ConstructionPolicyLaunchProfile {
  if (!isObject(payload)) {
    throw new Error('Construction policy catalog returned malformed launch_profile metadata')
  }
  if (payload.profile_id !== CANONICAL_LAUNCH_PROFILE_ID) {
    throw new Error('Construction policy catalog returned unsupported launch_profile.profile_id')
  }
  if (payload.profile_kind !== CANONICAL_LAUNCH_PROFILE_KIND) {
    throw new Error('Construction policy catalog returned unsupported launch_profile.profile_kind')
  }
  if (typeof payload.launch_top_n !== 'number' || !Number.isInteger(payload.launch_top_n)
    || payload.launch_top_n < RANKING_ARTIFACT_CONSTRUCTION_LAUNCH_TOP_N_MIN
    || payload.launch_top_n > RANKING_ARTIFACT_CONSTRUCTION_LAUNCH_TOP_N_MAX) {
    throw new Error('Construction policy catalog returned unsupported launch_profile.launch_top_n')
  }
  if (typeof payload.policy_status !== 'string' || !['default', 'opt_in', 'excluded'].includes(payload.policy_status)) {
    throw new Error('Construction policy catalog returned unsupported launch_profile.policy_status')
  }
  if (payload.policy_status !== CANONICAL_LAUNCH_POLICY_STATUS[policyId]) {
    throw new Error('Construction policy catalog returned policy metadata inconsistent with launch_profile.policy_status')
  }
  return {
    profile_id: CANONICAL_LAUNCH_PROFILE_ID,
    profile_kind: CANONICAL_LAUNCH_PROFILE_KIND,
    policy_status: payload.policy_status as ConstructionPolicyLaunchProfile['policy_status'],
    launch_top_n: payload.launch_top_n as number,
  }
}

function parseConstructionPolicyRow(payload: unknown): ConstructionDiscoveredPolicy {
  if (!isObject(payload)) {
    throw new Error('Construction policy catalog returned a malformed row')
  }
  const policyId = payload.policy_id
  if (typeof policyId !== 'string' || !policyId.trim() || !SUPPORTED_POLICY_IDS.has(policyId)) {
    throw new Error('Construction policy catalog returned an unsupported policy_id')
  }
  const expectedPolicySpec = SUPPORTED_POLICY_SPECS[policyId as keyof typeof SUPPORTED_POLICY_SPECS]
  const launchProfile = parseLaunchProfile(payload.launch_profile, policyId as keyof typeof SUPPORTED_POLICY_SPECS)
  const policyDefinitionId = payload.policy_definition_id
  if (typeof policyDefinitionId !== 'string' || policyDefinitionId !== expectedPolicySpec.policy_definition_id) {
    throw new Error('Construction policy catalog returned an unsupported policy_definition_id')
  }
  const name = payload.name
  if (typeof name !== 'string' || !name.trim()) {
    throw new Error('Construction policy catalog returned a malformed policy name')
  }
  const description = payload.description
  if (typeof description !== 'string' || !description.trim()) {
    throw new Error('Construction policy catalog returned a malformed policy description')
  }
  const family = payload.family
  if (typeof family !== 'string') {
    throw new Error('Construction policy catalog returned an unsupported policy family')
  }
  const rankingSupport = payload.ranking_support
  if (typeof rankingSupport !== 'string') {
    throw new Error('Construction policy catalog returned an unsupported ranking_support')
  }
  if (family !== expectedPolicySpec.family) {
    throw new Error('Construction policy catalog returned an unsupported policy family')
  }
  if (!SUPPORTED_RANKING_SUPPORT.has(rankingSupport)) {
    throw new Error('Construction policy catalog returned an unsupported ranking_support')
  }
  if (rankingSupport !== expectedPolicySpec.ranking_support) {
    throw new Error('Construction policy catalog returned policy metadata inconsistent with policy_id')
  }
  if (!Array.isArray(payload.selection_rule_ids) || payload.selection_rule_ids.length !== REQUIRED_SELECTION_RULE_IDS.length) {
    throw new Error('Construction policy catalog returned malformed selection_rule_ids')
  }
  const selectionRuleIds = payload.selection_rule_ids.map((value) => {
    if (typeof value !== 'string' || !value.trim()) {
      throw new Error('Construction policy catalog returned malformed selection_rule_ids')
    }
    return value
  })
  if (selectionRuleIds.some((value, index) => value !== REQUIRED_SELECTION_RULE_IDS[index])) {
    throw new Error('Construction policy catalog returned unsupported selection_rule_ids')
  }
  if (typeof payload.launch_top_n !== 'number' || !Number.isInteger(payload.launch_top_n)
    || payload.launch_top_n < RANKING_ARTIFACT_CONSTRUCTION_LAUNCH_TOP_N_MIN
    || payload.launch_top_n > RANKING_ARTIFACT_CONSTRUCTION_LAUNCH_TOP_N_MAX) {
    throw new Error('Construction policy catalog returned an unsupported launch_top_n')
  }
  if (payload.launch_top_n !== launchProfile.launch_top_n) {
    throw new Error('Construction policy catalog returned a row launch_top_n inconsistent with launch_profile.launch_top_n')
  }

  return {
    policy_id: policyId,
    policy_definition_id: policyDefinitionId,
    name,
    description,
    family,
    constraints: validateExactField(payload, 'constraints', DESKTOP_CONSTRUCTION_POLICY_FILTERS.constraints),
    inputs: validateExactField(payload, 'inputs', DESKTOP_CONSTRUCTION_POLICY_FILTERS.inputs),
    determinism: validateExactField(payload, 'determinism', DESKTOP_CONSTRUCTION_POLICY_FILTERS.determinism),
    ranking_support: rankingSupport as ConstructionDiscoveredPolicy['ranking_support'],
    full_investment_constraint: validateExactField(payload, 'full_investment_constraint', DESKTOP_CONSTRUCTION_POLICY_FILTERS.full_investment_constraint),
    long_only_constraint: validateExactField(payload, 'long_only_constraint', DESKTOP_CONSTRUCTION_POLICY_FILTERS.long_only_constraint),
    eligible_ranked_universe_constraint: validateExactField(payload, 'eligible_ranked_universe_constraint', DESKTOP_CONSTRUCTION_POLICY_FILTERS.eligible_ranked_universe_constraint),
    max_position_weight_constraint: validateExactField(payload, 'max_position_weight_constraint', DESKTOP_CONSTRUCTION_POLICY_FILTERS.max_position_weight_constraint),
    min_position_weight_constraint: validateExactField(payload, 'min_position_weight_constraint', DESKTOP_CONSTRUCTION_POLICY_FILTERS.min_position_weight_constraint),
    max_turnover_weight_constraint: validateExactField(payload, 'max_turnover_weight_constraint', DESKTOP_CONSTRUCTION_POLICY_FILTERS.max_turnover_weight_constraint),
    max_trade_intent_count_constraint: validateExactField(payload, 'max_trade_intent_count_constraint', DESKTOP_CONSTRUCTION_POLICY_FILTERS.max_trade_intent_count_constraint),
    ranked_universe_input: validateExactField(payload, 'ranked_universe_input', DESKTOP_CONSTRUCTION_POLICY_FILTERS.ranked_universe_input),
    current_portfolio_input: validateExactField(payload, 'current_portfolio_input', DESKTOP_CONSTRUCTION_POLICY_FILTERS.current_portfolio_input),
    launch_top_n: payload.launch_top_n as number,
    selection_rule_ids: selectionRuleIds,
    launch_profile: launchProfile,
  }
}

export function parseConstructionPolicyCatalog(payload: unknown): ConstructionDiscoveredPolicy[] {
  if (!Array.isArray(payload)) {
    throw new Error('Construction policy catalog returned a malformed payload')
  }
  const parsed = payload.map(parseConstructionPolicyRow)
  const seenPolicyIds = new Set<string>()
  const seenDefinitionIds = new Set<string>()
  for (const policy of parsed) {
    if (seenPolicyIds.has(policy.policy_id)) {
      throw new Error('Construction policy catalog returned duplicate policy_id rows')
    }
    if (seenDefinitionIds.has(policy.policy_definition_id)) {
      throw new Error('Construction policy catalog returned duplicate policy_definition_id rows')
    }
    seenPolicyIds.add(policy.policy_id)
    seenDefinitionIds.add(policy.policy_definition_id)
  }

  const profileRows = parsed.filter((policy) => policy.launch_profile.profile_id === CANONICAL_LAUNCH_PROFILE_ID)
  const defaultPolicies = profileRows.filter((policy) => policy.launch_profile.policy_status === 'default')
  if (defaultPolicies.length !== 1) {
    throw new Error('Construction policy catalog must define exactly one default launch policy')
  }
  if (defaultPolicies[0]?.policy_id !== 'top_n_equal_weight_v1') {
    throw new Error('Construction policy catalog returned unsupported default launch policy')
  }

  const includedPolicyIds = new Set(
    profileRows
      .filter(isLaunchIncludedPolicy)
      .map((policy) => policy.policy_id),
  )
  if (includedPolicyIds.size !== CANONICAL_LAUNCH_INCLUDED_POLICY_IDS.size) {
    throw new Error('Construction policy catalog returned unsupported launch-compatible policy set')
  }
  for (const policyId of CANONICAL_LAUNCH_INCLUDED_POLICY_IDS) {
    if (!includedPolicyIds.has(policyId)) {
      throw new Error('Construction policy catalog returned unsupported launch-compatible policy set')
    }
  }

  const inverseRankPolicy = parsed.find((policy) => policy.policy_id === 'top_n_inverse_rank_weight_v1')
  if (!inverseRankPolicy || inverseRankPolicy.launch_profile.policy_status !== 'excluded') {
    throw new Error('Construction policy catalog must exclude inverse-rank from the launch profile')
  }
  return parsed
}

export function buildConstructionPolicyCatalogUrl(apiBase = '/api') {
  const search = new URLSearchParams(DESKTOP_CONSTRUCTION_POLICY_FILTERS)
  return `${apiBase}/construction/policies?${search.toString()}`
}

export function getConstructionLaunchPolicyReadback(
  policyId: string | null,
  policies: ConstructionDiscoveredPolicy[],
): ConstructionLaunchPolicyReadback | null {
  if (!policyId) return null
  const policy = policies.find((entry) => entry.policy_id === policyId)
  if (!policy || !isLaunchIncludedPolicy(policy)) return null
  return {
    policyName: policy.name,
    statusLabel: policy.launch_profile.policy_status === 'default' ? 'default' : 'opt-in',
    topN: policy.launch_profile.launch_top_n,
    requiredConstraint: 'max_position_weight',
    optionalConstraints: ['min_position_weight', 'max_turnover_weight', 'max_trade_intent_count'],
  }
}

export function useConstructionPolicyCatalog(apiBase = '/api') {
  const [state, setState] = useState<ConstructionPolicyCatalogState>({
    status: 'idle',
    policies: [],
    error: null,
  })

  useEffect(() => {
    let active = true
    setState({ status: 'loading', policies: [], error: null })
    void (async () => {
      try {
        const response = await fetch(buildConstructionPolicyCatalogUrl(apiBase))
        const payload = await readJsonResponse<unknown>(response, 'Construction policy catalog is unavailable')
        const policies = parseConstructionPolicyCatalog(payload)
        if (!active) return
        setState({ status: 'ready', policies, error: null })
      } catch (caught) {
        if (!active) return
        setState({ status: 'error', policies: [], error: caught instanceof Error ? caught.message : 'Construction policy catalog is unavailable' })
      }
    })()
    return () => {
      active = false
    }
  }, [apiBase])

  const policies = useMemo(
    () => state.policies.filter(isLaunchIncludedPolicy),
    [state.policies],
  )

  const defaultPolicyId = useMemo(() => (
    policies.find((policy) => policy.launch_profile.policy_status === 'default')?.policy_id ?? null
  ), [policies])

  return {
    ...state,
    policies,
    defaultPolicyId,
  }
}

export function buildConstructionPolicyRunInput(policyId: string | null, topN: number): ConstructionPolicyRunInput | null {
  if (!policyId) return null
  if (!CANONICAL_LAUNCH_INCLUDED_POLICY_IDS.has(policyId)) return null
  if (!Number.isInteger(topN)
    || topN < RANKING_ARTIFACT_CONSTRUCTION_LAUNCH_TOP_N_MIN
    || topN > RANKING_ARTIFACT_CONSTRUCTION_LAUNCH_TOP_N_MAX) {
    throw new Error(
      `Ranking artifact construction launch requires integer top_n in `
      + `[${RANKING_ARTIFACT_CONSTRUCTION_LAUNCH_TOP_N_MIN}, ${RANKING_ARTIFACT_CONSTRUCTION_LAUNCH_TOP_N_MAX}] `
      + `(received ${topN})`
    )
  }
  return { policy_id: policyId, top_n: topN }
}
